code/components/gta-net-five/src/MumbleVoice.cpp

```diff
@@ -700,30 +700,119 @@ static HookFunction hookFunction([]()
 
 			if (id >= 0 && id < 31)
 			{
-				vtConfigs[id] = {};
-				g_mumbleClient->UpdateVoiceTarget(id, vtConfigs[id]);
+				if (g_mumble.connected)
+				{
+					vtConfigs[id] = {};
+					g_mumbleClient->UpdateVoiceTarget(id, vtConfigs[id]);
+				}
 			}
 		});
 
-		fx::ScriptEngine::RegisterNativeHandler("MUMBLE_CLEAR_VOICE_TARGET_CHANNELS", [](fx::ScriptContext& context)
+        fx::ScriptEngine::RegisterNativeHandler("MUMBLE_REMOVE_VOICE_TARGET_CHANNEL", [](fx::ScriptContext& context)
+        {
+            auto id = context.GetArgument<int>(0);
+            auto channel = context.GetArgument<int>(1);
+
+            if (id >= 0 && id < 31)
+            {
+				if (g_mumble.connected)
+				{
+					auto targetChannel = fmt::sprintf("Game Channel %d", channel);
+					auto& targets = vtConfigs[id].targets;
+					targets.remove_if([targetChannel](auto& target)
+					{
+						if (target.channel == targetChannel)
+						{
+							return true;
+						}
+						return false;
+					});
+
+					g_mumbleClient->UpdateVoiceTarget(id, vtConfigs[id]);
+				}
+            }
+        });
+
+		fx::ScriptEngine::RegisterNativeHandler("MUMBLE_REMOVE_VOICE_TARGET_PLAYER", [](fx::ScriptContext& context)
 		{
 			auto id = context.GetArgument<int>(0);
+			auto playerId = context.GetArgument<int>(1);
 
 			if (id >= 0 && id < 31)
 			{
-				std::vector<VoiceTargetConfig::Target>& targets = vtConfigs[id].targets;
-				for (size_t i = targets.size(); i--;)
+				if (g_mumble.connected)
 				{
-					VoiceTargetConfig::Target& target = targets[i];
-					if (!target.channel.empty())
+					std::wstring targetName = ToWide(fmt::sprintf("[%d] %s",
+					FxNativeInvoke::Invoke<int>(getServerId, playerId),
+					FxNativeInvoke::Invoke<const char*>(getPlayerName, playerId)));
+
+					auto& targets = vtConfigs[id].targets;
+					targets.remove_if([targetName](auto& target)
 					{
-						targets.erase(targets.begin() + i);
+						if (target.users.size() > 0 && std::find(target.users.begin(), target.users.end(), targetName) != target.users.end())
+						{
+							return true;
+						}
+						return false;
+					});
+
+					g_mumbleClient->UpdateVoiceTarget(id, vtConfigs[id]);
+				}
+			}
+		});
+
+		fx::ScriptEngine::RegisterNativeHandler("MUMBLE_REMOVE_VOICE_TARGET_PLAYER_BY_SERVER_ID", [](fx::ScriptContext& context)
+		{
+			auto id = context.GetArgument<int>(0);
+			auto serverId = context.GetArgument<int>(1);
+
+			if (id >= 0 && id < 31)
+			{
+				if (g_mumble.connected)
+				{
+					VoiceTargetConfig::Target ch;
+					std::wstring targetName = g_mumbleClient->GetPlayerNameFromServerId(serverId);
+
+					if (!targetName.empty())
+					{
+						auto& targets = vtConfigs[id].targets;
+						targets.remove_if([targetName](auto& target)
+						{
+							if (target.users.size() > 0 && std::find(target.users.begin(), target.users.end(), targetName) != target.users.end())
```

Comment:
size() > 0 check is a little unneeded, std::find handles begin == end fine

also, is each user a separate users target?

Reply:
This was left in because `MUMBLE_ADD_VOICE_TARGET_CHANNEL` also pushes onto vtConfigs[id].targets, so this was left as a fast denial

> also, is each user a separate users target?

I'm not entirely sure what you're asking